File: packages/parent-domain/src/lan-pairing-browser-runtime.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import { ParentDeviceIdSchema } from './reference-primitives';
import {
  LanPairingDiscoverySourceSchema,
  LanPairingOriginSchema,
  LanPairingRouteIdSchema,
  LanPairingSchemaVersionSchema,
  LanPairingTimestampSchema,
} from './lan-pairing-values';

export const LanBrowserDiscoveryScanRequestSchema = withParser(
  Schema.Struct({
    schemaVersion: LanPairingSchemaVersionSchema,
    requestedDiscoverySource: LanPairingDiscoverySourceSchema,
  })
);

export const LanBrowserAddDeviceRuntimeRequestSchema = withParser(
  Schema.Struct({
    schemaVersion: LanPairingSchemaVersionSchema,
    childDeviceId: ParentDeviceIdSchema,
    parentDeviceId: ParentDeviceIdSchema,
    routeId: LanPairingRouteIdSchema,
    origin: LanPairingOriginSchema,
    issuedAt: LanPairingTimestampSchema,
    expiresAt: LanPairingTimestampSchema,
  })
);

export type LanBrowserDiscoveryScanRequest = Infer<typeof LanBrowserDiscoveryScanRequestSchema>;
export type LanBrowserAddDeviceRuntimeRequest = Infer<typeof LanBrowserAddDeviceRuntimeRequestSchema>;

File: crates/agent-protocol/src/lan_pairing_browser_runtime.rs
use serde::{Deserialize, Serialize};

use crate::LanPairingDiscoverySource;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LanBrowserDiscoveryScanRequest {
    pub schema_version: u16,
    pub requested_discovery_source: LanPairingDiscoverySource,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LanBrowserAddDeviceRequest {
    pub schema_version: u16,
    pub child_device_id: String,
    pub parent_device_id: String,
    pub route_id: String,
    pub origin: String,
    pub issued_at: String,
    pub expires_at: String,
}
